docs/architecture/adr-064-batch-verification.md

```diff
@@ -0,0 +1,82 @@
+# ADR 064: Batch Verification
+
+## Changelog
+
+- January 28, 2021: Created (@marbar3778)
+
+## Context
+
+Tendermint uses public private key cryptography for validator signing. When a block is proposed and voted on validators sign a message representing acceptance or rejection of a block. These signatures are also used to verify previous blocks are correct if a node is syncing. Currently, Tendermint requires each signature to be verified individually, this leads to a slow down of block times.
+
+## Alternative Approaches
+
+- Signature aggregation
+  - Signature aggregation is an alternative to batch verification. Signature aggregation leads to fast verification and smaller block sizes. At the time of writing this ADR there is on going work to enable signature aggregation in Tendermint. The reason why we have opted to not introduce it at this time is because it will slow down verification times. One of the two benefits can be done, aggregation, but verification would suffer because the software would need to verify each signature individually due to their uniqueness, then aggregate them. For example if we were to implement signature aggregation with BLS, there could be a potential slow down of 10x-100x in verification speeds.
+
+## Decision
+
+Adopt Batch Verification.
+
+## Detailed Design
+
+A new interface will be introduced. This interface will have three methods `NewBatchVerifier`, `Add` and `VerifyBatch`.
+
+```go
+type BatchVerifier interface {
+  NewBatchVerifier() BatchVerifier // NewBatchVerifier create a new verifier where keys, signatures and messages can be added as entries
+  Add(key crypto.Pubkey, signature, message []byte) bool // Add appends an entry into the BatchVerifier.
+  Verify() bool // Verify verifies all the entries in the BatchVerifier. If the verification fails it is unknown which entry failed and each entry will need to be verified individually.
```

Comment:
nit: this may be OK in the Go context / within standard style conventions -- it's that a caller may mess up ("true" -- is it ok or failed? + one isn't forced to check errors; e.g. one may just iterate over pubkeys/sigs and call Add and throw away the returned bool)... returning `(...something..., error)` could make it more explicit (but `(struct{}, error)` is a bit ugly)

Reply:
I can change to return an error but if the verification fails the most it will say is "verification failed, verify all keys individually." Is that better than bool?

Since we dont know which key failed I didn't think adding a string like the above added much value but happy to change it to the string.

Reply:
go crypto returns `bool`: https://golang.org/pkg/crypto/ed25519/#Verify so perhaps better keep it consistent on `Verify`? For `Add`, it may be worth returning an error for compiler/lint warnings if one accidentally skips an invalid entry?

Reply:
amended add to return an error